mesonbuild/modules/selinux.py

```diff
@@ -0,0 +1,344 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2025 Red Hat, Inc
+
+from __future__ import annotations
+
+import os
+import glob
+import re
+import typing as T
+
+from . import ExtensionModule, ModuleReturnValue, ModuleInfo
+from .. import mesonlib
+from .. import build
+from .. import programs
+from ..build import CustomTarget
+from ..interpreter import Interpreter
+from ..interpreterbase.decorators import typed_kwargs, typed_pos_args
+from ..interpreter.type_checking import NoneType
+from ..mesonlib import File, MesonException
+from ..programs import ExternalProgram
+from ..interpreterbase.decorators import KwargInfo
+
+if T.TYPE_CHECKING:
+    from ..interpreterbase import TYPE_kwargs, TYPE_var
+    from . import ModuleState
+
+class SELinuxModule(ExtensionModule):
+    INFO = ModuleInfo('selinux', '1.9.0')
+
+    def __init__(self, interpreter: Interpreter) -> None:
+        super().__init__(interpreter)
+        self.methods.update({
+            'found': self.found_method,
+            'package': self.package,
+        })
+        self._tools: T.Optional[T.Dict[str, ExternalProgram]] = None
+
+    def _find_tools(self, state: ModuleState) -> T.Dict[str, ExternalProgram]:
+        if self._tools is None:
+            self._tools = {
+                'm4': T.cast(ExternalProgram, state.find_program('m4', required=False)),
+                'checkmodule': T.cast(ExternalProgram, state.find_program('checkmodule', required=False)),
+                'semodule_package': T.cast(ExternalProgram, state.find_program('semodule_package', required=False)),
+            }
+        return self._tools
+
+    def found_method(self, state: ModuleState, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> bool:
+        return all(t.found() for t in self._find_tools(state).values() if t)
+
+    def _ensure_file(self, state: ModuleState, file: T.Union[str, File], default_content: T.Optional[str] = None) -> T.Union[File, build.CustomTarget]:
+        sourcedir = state.environment.get_source_dir()
+        subdir = state.subdir
+
+        if isinstance(file, str):
+            if os.path.exists(os.path.join(state.environment.source_dir, subdir, file)) or default_content is None:
+                return File.from_source_file(sourcedir, subdir, file)
+
+            if default_content:
+                py_cmd = f"with open('@OUTPUT@', 'w') as f: f.write('{default_content}')"
+                desc = f"Creating file {file} with content"
```

Comment:
This is broken in a variety of different ways. Leaving aside the issue that you're producing a python script that writes static output at build time instead of just... creating static configure-time content ***at configure time***:
- it's unidiomatic to write "python -c program_code" as part of the meson core when you can instead write a script in mesonbuild/scripts/*.py and invoke it via `meson --internal scriptname` generally resulting in far shorter command lines that don't have to handle quoting and escaping hell for syntactic whitespace, newlines, backslash characters and more
- if either default_content or `@OUTPUT@` contain single quote characters this will explode

Reply:
I made changes so that it's all handled during build now (which has some pros and cons). The intermediate temporary files are build details, I don't think we should create them during configure time (we could though).